Add ability to render bar charts and legends for bar charts

dougnd opened the pull request:
Hello again.  This PR is a bit more complicated and may have to be modified before you are likely to accept it.

I was looking to add the ability to render bar charts.   Even with draw_rectangles turned on, I had Latex errors, and even if that worked, the legend would not be drawn appropriately.

There are two major changes here:
1. Rectangles with a label are draw (ignoring draw_rectangles). It seems to me, rectangles with labels (although they are typically set to '_nolegend_' ) are associated with a plot and should be drawn.
2. Since bar charts have no \addplot associated, a legend must be added manually with \addlegendimage 
  From page 249 of the PGFPlots manual: 

    > Use \addlegendimage to provide custom styles into legends, for example to document custom \draw
commands inside of an axis.
 
    So, I set it up so that the first rectangle of each label has an \addlegendimage added

One confusing part is getting the legend label associated with the rectangle.  For that, see  [this](http://stackoverflow.com/questions/35881290/how-to-get-the-label-on-bar-plot-stacked-bar-plot-in-matplotlib) discussion.

The barchart test output (showing working legends):
![barchartjtffuz-1](https://cloud.githubusercontent.com/assets/5289854/22603110/0ef97d0c-ea14-11e6-81f1-1af8d1088bc9.png)

The reference:
![reference](https://cloud.githubusercontent.com/assets/5289854/22603179/4cef247c-ea14-11e6-980c-5e24b47441d9.png)

nschloe commented:
Very nice!

With `data['draw rectangles']` disregarded, this option could be scratched from `save()` altogether.

Does this PR solve #126 by any chance?

dougnd commented:
Yep, that seems to fix #126. I get this as an output:

![issue1264uopek-1](https://cloud.githubusercontent.com/assets/5289854/22604296/dc5c499c-ea18-11e6-882e-381bfc2a3f32.png)


I was not sure what to do with `draw_rectangles`.  Would you like me to strip it out altogether? Leave it in but use a warning message?  Or I could even change
```
if label == '':
          return data, []
```
to
```
if label == '' and not data['draw rectangles']:
          return data, []
```
on line [61 of patch.py](https://github.com/nschloe/matplotlib2tikz/pull/135/files#diff-56d3c0a4ef8fd66f7b4cc04edb2ee20eR61) if there was still some reason to force all rectangles to be drawn.

nschloe commented:
> I was not sure what to do with `draw_rectangles`.

That variable is a more desperation than anything useful. Now that we can separate good from bad boxes (at least a little bit, by only considering the labeled ones), we don't really need it anymore. I'd say if your plot above works without labels too, then get rid of `draw_rectangles`. Perhaps a test for that would be in order too (in case we don't have one yet, can't remember).

dougnd commented:
Ok cool.  I've removed all mention of `draw_rectangles`.  This does include a use in test_hashes.py (see the diff).  I'm not sure what "test other height specs" does, but I'm assuming that removing `draw_rectangles` should be ok.

I've also added a barchart test with no legend/labels and it seems to behave correctly:

![barcharte5hiwk-1](https://cloud.githubusercontent.com/assets/5289854/22605445/e220b962-ea1d-11e6-9394-e75883213332.png)

dougnd commented:
One thing I should note before you merge this is that I use the legend style of `ybar,ybar legend` for any labeled rectangle plots (including histograms and horizontal bar charts).  In the future maybe that should be changed.

nschloe commented:
Very nice work again, thank you!
